src/analysis/convex/cone/proper.lean

```diff
@@ -46,8 +47,89 @@ protected def closure (K : convex_cone ℝ E) : convex_cone ℝ E :=
 
 @[simp, norm_cast] lemma coe_closure (K : convex_cone ℝ E) : (K.closure : set E) = closure K := rfl
 
-protected lemma mem_closure {K : convex_cone ℝ E} {a : E} :
-  a ∈ K.closure ↔ a ∈ closure (K : set E) :=
-iff.rfl
+@[simp] protected lemma mem_closure {K : convex_cone ℝ E} {a : E} :
+  a ∈ K.closure ↔ a ∈ closure (K : set E) := iff.rfl
 
 end convex_cone
+
+section definitions
+
+/-- A proper cone is a convex cone `K` that is nonempty and closed. Proper cones have the nice
+property that the dual of the dual of a proper cone is itself. This makes them useful for defining
+cone programs and proving duality theorems. -/
+structure proper_cone (E : Type*) [normed_add_comm_group E] [inner_product_space ℝ E]
+  extends convex_cone ℝ E :=
+(nonempty'  : (carrier : set E).nonempty)
+(is_closed' : is_closed (carrier : set E))
+
+end definitions
+
+namespace proper_cone
+
+section inner_product_space
+
+variables {E : Type*} [normed_add_comm_group E] [inner_product_space ℝ E]
+variables {F : Type*} [normed_add_comm_group F] [inner_product_space ℝ F]
+
+instance : has_coe (proper_cone E) (convex_cone ℝ E) := ⟨λ K, K.1⟩
+
+instance : has_mem E (proper_cone E) := ⟨λ e K, e ∈ (K : convex_cone ℝ E) ⟩
+
+noncomputable instance : has_zero (proper_cone E) := ⟨ ⟨0, ⟨0, rfl⟩, is_closed_singleton⟩ ⟩
+
+noncomputable instance : inhabited (proper_cone E) := ⟨0⟩
+
+@[simp] lemma mem_coe {x : E} {K : proper_cone E} : x ∈ K ↔ x ∈ (K : convex_cone ℝ E) := iff.rfl
+
+protected lemma nonempty (K : proper_cone E) : (K : set E).nonempty := K.nonempty'
+
+protected lemma is_closed (K : proper_cone E) : is_closed (K : set E) := K.is_closed'
+
+protected lemma pointed (K : proper_cone E) : (K : convex_cone ℝ E).pointed :=
+(K : convex_cone ℝ E).pointed_of_nonempty_of_is_closed K.nonempty K.is_closed
+
+@[ext] lemma ext {S T : proper_cone E} (h : (S : convex_cone ℝ E) = T) : S = T :=
+by cases S; cases T; congr'
+
+/-- We define `K*` to be the inner dual cone of `K`. -/
+instance : has_star (proper_cone E) :=
```

Comment:
This strikes me as an abuse of the `has_star` typeclass; that's intended primarily for C-star algebras, or modules over them. As far as I can tell there's no connection here

Reply:
I see. I'll get rid of this instance and just have a lemma about involutive dual.